compiler/rustc_infer/src/infer/error_reporting/suggest.rs

```diff
@@ -351,6 +351,73 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
         }
     }
 
+    pub(super) fn suggest_function_pointers(
+        &self,
+        exp_found: &ty::error::ExpectedFound<Ty<'tcx>>,
+        diag: &mut Diagnostic,
+    ) {
+        let ty::error::ExpectedFound { expected, found } = exp_found;
+        if expected.peel_refs().is_fn() && found.peel_refs().is_fn() {
+            let expected_inner = expected.peel_refs();
+            let found_inner = found.peel_refs();
+            diag.note(
+                "different `fn` items always have unique types, even if their signatures are \
+                     the same",
+            );
+            match (&expected_inner.kind(), &found_inner.kind()) {
+                (ty::FnDef(did1, substs1), ty::FnDef(did2, substs2)) => {
+                    let sig1 = self.tcx.bound_fn_sig(*did1).subst(self.tcx, substs1);
+                    let sig2 = self.tcx.bound_fn_sig(*did2).subst(self.tcx, substs2);
+                    if sig1 != sig2 {
+                        return;
+                    }
```

Comment:
This is in `TypeErrCtxt`, right? You could probably pass both of these signatures thru `(self.normalize_fn_sig)(fn_sig)`, like `TypeErrCtxt::cmp_fn_sig` does.